src/metabase/csv.clj

```diff
@@ -0,0 +1,96 @@
+(ns metabase.csv
+  (:require
+   [clojure.data.csv :as csv]
+   [clojure.java.io :as io]
+   [clojure.string :as str]
+   [medley.core :as m]
+   [metabase.search.util :as search-util]
+   [metabase.util :as u]
+   [metabase.util.i18n :refer [trs]]))
+
+;;           text
+;;            |
+;;            |
+;;       varchar_255
+;;            |
+;;            |
+;;          float
+;;            |
+;;            |
+;;           int
+;;            |
+;;            |
+;;         boolean
+(derive ::boolean     ::int)
+(derive ::int         ::float)
+(derive ::float       ::varchar_255)
+(derive ::varchar_255 ::text)
+
+(defn value->type
+  "The most-specific possible type for a given value. Possibilities are:
+    - ::boolean
+    - ::int
+    - ::float
+    - ::varchar_255
+    - ::text
+    - nil, in which case other functions are expected to replace it with ::text as the catch-all type
+
+  NB: There are currently the following gotchas:
+    1. ints/floats are assumed to have commas as separators and periods as decimal points
+    2. 0 and 1 are assumed to be booleans, not ints."
+  [value]
+  (cond
+    (str/blank? value)                                      nil
+    (re-matches #"(?i)true|t|yes|y|1|false|f|no|n|0" value) ::boolean
+    (re-matches #"-?[\d,]+"                          value) ::int
+    (re-matches #"-?[\d,]*\.\d+"                     value) ::float
+    (re-matches #".{1,255}"                          value) ::varchar_255
+    :else                                                   ::text))
+
+(defn- row->types
+  [column-names row]
+  (->> row
+       (map vector column-names)
+       (map (fn [[column-name value]] [column-name (and value (value->type (search-util/normalize value)))]))
+       (into {})))
+
+(defn coalesce
+  "Returns the 'parent' type (the most general)."
+  [type-a type-b]
+  (cond
+    (nil? type-a)        type-b
+    (nil? type-b)        type-a
+    (isa? type-a type-b) type-b
+    (isa? type-b type-a) type-a
+    :else (throw (Exception. (trs "Unexpected type combination in the same column: {0} and {1}" type-a type-b)))))
+
+(defn- coalesce-types
+  [types-so-far new-types]
+  (->> types-so-far
+       (map (fn [[column-name old-type]]
+              [column-name (coalesce old-type (get new-types column-name))]))
+       (into {})))
+
+(defn- rows->schema
+  [header rows]
+  (let [normalized-header (map (comp u/slugify str/trim) header)]
+    (->> rows
+         (map (partial row->types normalized-header))
+         (reduce coalesce-types)
+         (m/map-vals #(or % ::text)))))
+
+(defn detect-schema
+  "Returns a map of `normalized-column-name -> type` for the given CSV file. The CSV file *must* have headers as the
```

Comment:
To create the table, the schema needs to include the ordering of the columns in the csv. It `detect-schema` should return a list of maps, where each map includes the type and normalized-column-name. The position of each map in the list should match the position of the column in the csv row.